### PR TITLE
Support nested input columns in copy_if_else()

### PR DESCRIPTION
**Pull request by mythrocks**
This commit aims to extend `copy_if_else()` to work with `LIST` and `STRUCT` inputs.

In this first version, `LHS` and `RHS` have to be `column_view`. Support for nested scalar operands can be added at a later date.

### CODE REVIEW

**Comment by mythrocks** [comment]
This initial draft implementation is not yet ready for review. For the moment, it's up for @hyperbolic2346's perusal.

**Comment by mythrocks** [comment]
This PR depends on #8150 being fixed.

**Comment by mythrocks** [comment]
Ok, I'm now running into incorrectly set validity masks in `STRUCT` children, in the output of `scatter()`. I have filed #8162 to track this.

I'm changing the scope of the current PR to address only `LIST` inputs. I will deal with `STRUCT` separately and afterwards.

File: cpp/src/copying/copy.cu [hunk 1]
```diff
@@ -38,18 +40,40 @@ struct copy_if_else_functor_impl {
   }
 };
 
+/**
+ * @brief Functor to fetch a device-view for the specified scalar/column_view.
```
Comment (hyperbolic2346) [inline cpp/src/copying/copy.cu:44]: Would `constexpr if` not work for this? I'm mostly curious if it failed more than trying to push you to it.

File: cpp/src/copying/copy.cu [hunk 2]
```diff
@@ -38,18 +40,40 @@ struct copy_if_else_functor_impl {
...
                                      size_type size,
                                      bool left_nullable,
                                      bool right_nullable,
                                      Filter filter,
                                      rmm::cuda_stream_view stream,
                                      rmm::mr::device_memory_resource* mr)
   {
+    auto p_lhs      = get_iterable_device_view{}(lhs_h);
+    auto p_rhs      = get_iterable_device_view{}(rhs_h);
+    auto const& lhs = *p_lhs;
```
Comment (hyperbolic2346) [inline cpp/src/copying/copy.cu:74]: If there isn't a nullptr check, why break this into multiple lines?

File: cpp/src/copying/copy.cu [hunk 3]
```diff
@@ -114,6 +143,125 @@ struct copy_if_else_functor_impl<string_view> {
...
+          CUDF_ENABLE_IF(not std::is_same<Left, cudf::scalar>::value &&
+                         not std::is_same<Right, cudf::scalar>::value)>
+std::unique_ptr<column> scatter_gather_based_if_else(Left const& lhs,
+                                                     Right const& rhs,
+                                                     size_type size,
+                                                     Filter is_left,
+                                                     rmm::cuda_stream_view stream,
+                                                     rmm::mr::device_memory_resource* mr)
+{
+  auto null = size + 1;  // Out of bounds index, for gather() to nullify.
```
Comment (hyperbolic2346) [inline cpp/src/copying/copy.cu:226]:
```suggestion
  auto const null_map_entry = size + 1;  // Out of bounds index, for gather() to nullify.
```

I'm nervous about overloading the name null in general.

**Review by hyperbolic2346 — COMMENTED** [review]
Very nice. I think the functors may be a bit less readable than a lambda, but it's also possible that compiler issues/features prevent this. I'm fine either way. This would be one of the places I would like to see a large comment at the top explaining how `gather` and `scatter` result in a `copy_if_else`. Not sure we can get that in though.

**Comment by mythrocks** [comment]
> Very nice. I think the functors may be a bit less readable than a lambda, but it's also possible that compiler issues/features prevent this.

These started off as lambdas with captures. And then I ran into the `_T201 ...` errors from `nvcc`. I had to switch to using functors instead. :/

File: cpp/src/copying/copy.cu (hunk 1 above)
Comment (mythrocks) [reply cpp/src/copying/copy.cu:44]: I haven't been able to boil this down to something reproducible via `godbolt`, to illustrate the problem. I can try again.

Comment (hyperbolic2346) [reply cpp/src/copying/copy.cu:44]: If it won't work, that's fine.

File: cpp/src/copying/copy.cu (hunk 3 above)
Comment (mythrocks) [reply cpp/src/copying/copy.cu]: Fixed.

File: cpp/src/copying/copy.cu (hunk 2 above)
Comment (mythrocks) [reply cpp/src/copying/copy.cu:74]: `get_iterable_device_view()` returns a `unique_ptr<column_device_view>`. It needs to be kept alive, and can't be dereferenced directly.

File: cpp/src/copying/copy.cu (hunk 1 above)
Comment (mythrocks) [reply cpp/src/copying/copy.cu:44]: I'm taking another crack at this.

Comment (mythrocks) [reply cpp/src/copying/copy.cu:44]: Nope. I can't seem to get around the warning. For the moment, I'll leave this where it is.

**Comment by mythrocks** [comment]
Note: #8176 looks like it fixes the `struct` problems I've been facing in this PR. Once that is merged, I'll add the `struct` tests back in, and enable the `struct` code path.

File: cpp/src/copying/copy.cu (hunk 1 above)
Comment (ttnghia) [reply cpp/src/copying/copy.cu:44]: The two `operator()` return different types, thus you can't use `constexpr if`.

File: cpp/src/copying/copy.cu [hunk 4]
```diff
@@ -114,6 +143,125 @@ struct copy_if_else_functor_impl<string_view> {
...
+          CUDF_ENABLE_IF(not std::is_same<Left, cudf::scalar>::value &&
+                         not std::is_same<Right, cudf::scalar>::value)>
+std::unique_ptr<column> scatter_gather_based_if_else(Left const& lhs,
+                                                     Right const& rhs,
+                                                     size_type size,
+                                                     Filter is_left,
+                                                     rmm::cuda_stream_view stream,
+                                                     rmm::mr::device_memory_resource* mr)
+{
+  auto null_map_entry = size + 1;  // Out of bounds index, for gather() to nullify.
```
Comment (nvdbaranec) [inline cpp/src/copying/copy.cu:226]:
```suggestion
  auto const null_map_entry = size + 1;  // Out of bounds index, for gather() to nullify.
```

File: cpp/src/copying/copy.cu [hunk 5]
```diff
@@ -114,6 +143,125 @@ struct copy_if_else_functor_impl<string_view> {
...
+                                         logical_not{is_left});
+
+  auto scatter_src_rhs = cudf::detail::gather(table_view{std::vector<column_view>{rhs}},
+                                              scatter_map_rhs.begin(),
+                                              scatter_map_end,
+                                              out_of_bounds_policy::DONT_CHECK,
+                                              stream);
+
+  auto result = cudf::detail::scatter(
+    table_view{std::vector<column_view>{scatter_src_rhs->release()[0]->view()}},
```
Comment (nvdbaranec) [inline cpp/src/copying/copy.cu:254]: This seems dangerous.  
```suggestion
    table_view{std::vector<column_view>{scatter_src_rhs->get_column(0)->view()}},
```

**Review by nvdbaranec — CHANGES_REQUESTED** [review]
Note:  it looks like 8162 was fixed with   https://github.com/rapidsai/cudf/pull/8176

File: cpp/src/copying/copy.cu (hunk 5 above)
Comment (mythrocks) [reply cpp/src/copying/copy.cu:254]: Hmm... It isn't, though: the `vector<unique_ptr<>>` should be "alive" as long the expression in which it's defined is "alive".

That said, your way is far more preferable: the behaviour is unambiguous, and the fetch of the column is explicit. 

Thanks for pointing this out. I'll make this change very shortly.

Comment (mythrocks) [reply cpp/src/copying/copy.cu]: Fixed.

File: cpp/src/copying/copy.cu (hunk 4 above)
Comment (mythrocks) [reply cpp/src/copying/copy.cu]: Fixed. `const`ed all the things.

File: cpp/src/copying/copy.cu (hunk 3 above)
Comment (mythrocks) [reply cpp/src/copying/copy.cu]: Yikes. I'd forgotten the `const` when I resolved this. @nvdbaranec caught it also.

File: cpp/src/copying/copy.cu (hunk 1 above)
Comment (mythrocks) [reply cpp/src/copying/copy.cu:44]: > The two `operator()` return different types, thus you can't use `constexpr if`.

@ttnghia, I'm not sure which `operator()` you're referring to here, but I have switched over to using `if constexpr`. 

@hyperbolic2346: The latest commit has one way to work around the warning. (Hat tip to @vuule for the advice.)

**Review by nvdbaranec — APPROVED** [review]

**Comment by mythrocks** [comment]
Thanks for the reviews, @ttnghia, @nvdbaranec. 
@hyperbolic2346, does this look agreeable?

**Review by hyperbolic2346 — APPROVED** [review]

**Comment by mythrocks** [comment]
@gpucibot merge

**Pull request merged**